File: src/level.py
"""
Level class - Handles platforms and level layout
"""
import pygame
from config import *

class Level:
    """Game level with platforms and decorations"""
    
    def __init__(self, screen_width, screen_height):
        self.width = 3000  # Total level width
        self.height = screen_height
        self.platforms = []
        self.boundaries = []  # Invisible walls
        
        # Background elements (clouds, mountains)
        self.clouds = self.create_clouds()
        self.mountains = self.create_mountains()
        
        # Create platforms
        self.create_platforms()
        self.create_boundaries()
    
    def create_platforms(self):
        """Create platform layout with proper clearance (80px vertical, 80px horizontal min)"""
        # Ground platform (full width)
        self.platforms.append(pygame.Rect(0, 580, self.width, 40))
        
        # Starting area platforms - proper spacing (player is 64px wide/tall)
        self.platforms.append(pygame.Rect(200, 480, 150, 20))  # 100px clearance above
        self.platforms.append(pygame.Rect(430, 420, 150, 20))  # 80px gap from prev, 160px clearance above
        
        # Mid-section with varied heights - ensuring passable gaps
        self.platforms.append(pygame.Rect(680, 460, 180, 20))  # 100px gap, 120px clearance
        self.platforms.append(pygame.Rect(960, 380, 150, 20))  # 100px gap, 200px clearance - high platform
        self.platforms.append(pygame.Rect(1210, 440, 150, 20))  # 100px gap, 140px clearance
        self.platforms.append(pygame.Rect(1460, 360, 180, 20))  # 100px gap, 220px clearance - high platform
        
        # High platforms section - accessible jumps with ground passage
        self.platforms.append(pygame.Rect(1740, 300, 150, 20))  # 100px gap, 280px clearance - high platform
        self.platforms.append(pygame.Rect(1990, 360, 140, 20))  # 100px gap, 220px clearance
        
        # Raised floor sections - blocking ground passage intentionally
        self.platforms.append(pygame.Rect(2230, 540, 180, 40))  # 100px gap, raised floor at 540
        self.platforms.append(pygame.Rect(2510, 500, 180, 40))  # 100px gap, raised floor at 500
        
        # Staircase on raised sections
        self.platforms.append(pygame.Rect(2790, 430, 120, 20))  # 100px gap, 150px clearance
        
        # Final area - accessible from staircase with gap
        self.platforms.append(pygame.Rect(3000, 400, 0, 20))  # At boundary (no actual platform needed here)
    
    def create_clouds(self):
        """Create parallax clouds across the level"""
        clouds = []
        # Layer 1 - Far clouds (slower parallax)
        for i in range(8):
            x = i * 400 + 100
            y = 80 + (i % 3) * 30
            width = 80 + (i % 2) * 40
            height = 40 + (i % 2) * 15
            clouds.append({'x': x, 'y': y, 'width': width, 'height': height, 'layer': 1})
        
        # Layer 2 - Near clouds (faster parallax)
        for i in range(10):
            x = i * 350 + 200
            y = 120 + (i % 4) * 25
            width = 100 + (i % 3) * 30
            height = 50 + (i % 2) * 10
            clouds.append({'x': x, 'y': y, 'width': width, 'height': height, 'layer': 2})
        
        return clouds
    
    def create_mountains(self):
        """Create distant mountain silhouettes"""
        mountains = []
        # Create 5 mountain ranges
        for i in range(5):
            x = i * 650
            base_y = 400
            height = 180 + (i % 3) * 60
            width = 500 + (i % 2) * 150
            mountains.append({'x': x, 'y': base_y, 'width': width, 'height': height})
        
        return mountains
    
    def create_boundaries(self):
        """Create invisible walls at level edges"""
        # Left wall
        self.boundaries.append(pygame.Rect(-50, 0, 50, self.height))
        
        # Right wall
        self.boundaries.append(pygame.Rect(self.width, 0, 50, self.height))
        
        # Death zone below level
        self.boundaries.append(pygame.Rect(0, 650, self.width, 50))
    
    def check_collision(self, rect, velocity_y):
        """Check if rect collides with platforms"""
        for platform in self.platforms:
            if rect.colliderect(platform):
                if velocity_y > 0:  # Falling
                    return True, platform.top
        return False, 0
    
    def render(self, screen, camera_x):
        """Render the level"""
        # Cyberpunk sky gradient - dark purple to magenta
        for y in range(0, 400, 20):
            ratio = y / 400
            # Deep purple at top to hot magenta at horizon
            r = int(25 + (138 - 25) * ratio)
            g = int(0 + (43 - 0) * ratio)
            b = int(51 + (226 - 51) * ratio)
            pygame.draw.rect(screen, (r, g, b), (0, y, SCREEN_WIDTH, 20))
        
        # Draw distant mountains (parallax layer 0.2x)
        for mountain in self.mountains:
            screen_x = mountain['x'] - (camera_x * 0.2)
            if -mountain['width'] < screen_x < SCREEN_WIDTH:
                # Mountain peak as triangle
                peak_x = screen_x + mountain['width'] // 2
                peak_y = mountain['y'] - mountain['height']
                base_left = screen_x
                base_right = screen_x + mountain['width']
                base_y = mountain['y']
                
                # Mountain silhouette (dark cyan-purple)
                points = [(peak_x, peak_y), (base_left, base_y), (base_right, base_y)]
                pygame.draw.polygon(screen, (20, 30, 60), points)
                
                # Neon cyan peak glow (top 20% of mountain)
                snow_height = mountain['height'] * 0.2
                snow_left_x = peak_x - snow_height * 0.5
                snow_right_x = peak_x + snow_height * 0.5
                snow_y = peak_y + snow_height
                snow_points = [(peak_x, peak_y), (snow_left_x, snow_y), (snow_right_x, snow_y)]
                pygame.draw.polygon(screen, (0, 255, 255), snow_points)
        
        # Draw clouds with parallax
        for cloud in self.clouds:
            # Different parallax speeds for different layers
            parallax = 0.3 if cloud['layer'] == 1 else 0.5
            screen_x = cloud['x'] - (camera_x * parallax)
            
            if -cloud['width'] < screen_x < SCREEN_WIDTH:
                # Draw clouds as neon pink/purple vapor
                color = (255, 0, 200) if cloud['layer'] == 1 else (200, 0, 255)
                # Main cloud body
                pygame.draw.ellipse(screen, color, 
                                  (screen_x, cloud['y'], cloud['width'], cloud['height']))
                # Additional puffs for depth
                pygame.draw.ellipse(screen, color, 
                                  (screen_x + cloud['width'] * 0.2, cloud['y'] - cloud['height'] * 0.2, 
                                   cloud['width'] * 0.5, cloud['height'] * 0.8))
                pygame.draw.ellipse(screen, color, 
                                  (screen_x + cloud['width'] * 0.5, cloud['y'] - cloud['height'] * 0.15, 
                                   cloud['width'] * 0.6, cloud['height'] * 0.9))
        
        # Draw platforms
        for platform in self.platforms:
            screen_x = platform.x - camera_x
            
            # Different colors for ground vs floating platforms
            if platform.y >= 580:
                # Ground - neon grid floor
                base_dark = (10, 10, 30)  # Very dark blue
                neon_cyan = (0, 255, 255)  # Bright cyan
                
                # Draw dark base
                pygame.draw.rect(screen, base_dark,
                               (screen_x, platform.y + 8, platform.width, platform.height - 8))
                
                # Draw cyan energy layer
                pygame.draw.rect(screen, (0, 100, 120),
                               (screen_x, platform.y, platform.width, 8))
                
                # Add neon grid lines
                for i in range(0, platform.width, 8):
                    # Draw vertical neon lines
                    blade_x = screen_x + i
                    pygame.draw.line(screen, neon_cyan, 
                                   (blade_x + 2, platform.y + 7), 
                                   (blade_x + 2, platform.y + 2), 2)
                    pygame.draw.line(screen, neon_cyan,
                                   (blade_x + 5, platform.y + 7),
                                   (blade_x + 5, platform.y + 3), 2)
                
                # Bright cyan top edge
                pygame.draw.rect(screen, neon_cyan,
                               (screen_x, platform.y, platform.width, 2))
            else:
                # Floating platforms - holographic purple/magenta
                base_purple = (60, 20, 80)
                dark_purple = (30, 10, 50)
                neon_magenta = (255, 0, 255)
                
                # Main platform body
                pygame.draw.rect(screen, base_purple,
                               (screen_x, platform.y, platform.width, platform.height))
                
                # Neon magenta edge on top
                pygame.draw.rect(screen, neon_magenta,
                               (screen_x, platform.y, platform.width, 3))
                
                # Dark shadow/depth
                if platform.height > 10:
                    pygame.draw.rect(screen, dark_purple,
                                   (screen_x, platform.y + 3, platform.width, platform.height - 3))
                
                # Bright magenta highlight
                highlight_color = (200, 50, 255)
                pygame.draw.rect(screen, highlight_color,
                               (screen_x, platform.y, platform.width, 2))
            
            # Platform outline - neon cyan glow
            pygame.draw.rect(screen, (0, 255, 255),
                           (screen_x, platform.y, platform.width, platform.height), 2)
        
        # Draw boundaries (visual indicators)
        for boundary in self.boundaries:
            screen_x = int(boundary.x - camera_x)
            
            # Only draw if visible on screen
            if -100 < screen_x < screen.get_width() + 100:
                # Left wall (neon magenta barrier)
                if boundary.x < 0:
                    # Draw warning stripes
                    for i in range(0, self.height, 40):
                        color = (255, 0, 150) if (i // 40) % 2 == 0 else (150, 0, 100)
                        pygame.draw.rect(screen, color,
                                       (screen_x, i, 50, 40))
                    # Outline
                    pygame.draw.rect(screen, (255, 0, 255),
                                   (screen_x, 0, 50, self.height), 3)
                
                # Right wall (neon magenta barrier)
                elif boundary.x >= self.width:
                    # Draw warning stripes
                    for i in range(0, self.height, 40):
                        color = (255, 0, 150) if (i // 40) % 2 == 0 else (150, 0, 100)
                        pygame.draw.rect(screen, color,
                                       (screen_x, i, 50, 40))
                    # Outline
                    pygame.draw.rect(screen, (255, 0, 255),
                                   (screen_x, 0, 50, self.height), 3)

File: js/level.js
class Level {
    /**
     * Load level data from an external configuration object.
     * Accepts partial data; missing fields will keep previous values.
     * @param {Object} levelData
     */
    loadLevel(levelData = {}) {
        this.width = levelData.width || this.width;
        this.height = levelData.height || this.height;
        // Optional background name (matches keys from spriteLoader)
        this.backgroundName = levelData.background || levelData.backgroundName || this.backgroundName || 'bg_city';
        // Additional background layers
        this.backgroundLayers = Array.isArray(levelData.backgroundLayers) ? levelData.backgroundLayers.map(layer => ({ ...layer })) : [];
        // Lazy-load background panoramas (non-blocking). Construct a simple
        // path mapping from sprite key to expected asset path and request
        // the spriteLoader to fetch them only when a level references them.
        try {
            if (typeof spriteLoader !== 'undefined') {
                const ensureLoaded = (name) => {
                    if (!name) return;
                    if (spriteLoader.getSprite(name)) return; // already present
                    // map bg_xxx -> assets/sprites/backgrounds/xxx_bg
                    const base = name.replace(/^bg_/, '');
                    const basePath = `assets/sprites/backgrounds/${base}_bg`;
                    // Use the robust variant loader which tries webp/png and suffixes
                    try {
                        spriteLoader.loadSpriteBest(name, basePath).then(img => { try { this.cachedSprites[name] = img; } catch (e) {} }).catch(() => {});
                    } catch (e) {}
                };

                ensureLoaded(this.backgroundName);
                for (const layer of this.backgroundLayers) {
                    if (layer && layer.name) ensureLoaded(layer.name);
                }
            }
        } catch (e) {}
        // Per-level background parallax factor (0..1). Lower = slower (farther away).
        this.backgroundParallax = (typeof levelData.backgroundParallax !== 'undefined') ? levelData.backgroundParallax : (typeof Config !== 'undefined' ? Config.BACKGROUND_PARALLAX : 0.5);
        
        // Initialize platforms (static and moving)
        // Accept partial data: if platforms is missing, keep existing platforms.
        const incomingPlatforms = Array.isArray(levelData.platforms) ? levelData.platforms : this.platforms;
        this.platforms = (Array.isArray(incomingPlatforms) ? incomingPlatforms : []).map(p => ({
            ...p,
            // If it's a moving platform, set initial state
            initialX: p.x,
            initialY: p.y,
            timeOffset: Math.random() * Math.PI * 2 // Randomize start phase
        }));
        // Optional enemy spawn points (array of { x: number|'left'|'right', y: number })
        this.spawnPoints = Array.isArray(levelData.spawnPoints) ? levelData.spawnPoints.slice() : null;
        
        // Store boss and completion config for Game logic
        this.bossConfig = levelData.boss || null;
        this.completionConfig = levelData.completion || null;

        // Mark static layer dirty so it will be re-rendered on next draw
        this._staticNeedsUpdate = true;
        // Hazards removed: clear all hazards on load to prevent spawning
        const hadHazards = Array.isArray(levelData.hazards) && levelData.hazards.length > 0;
        this.hazards = [];
        if (hadHazards && typeof Config !== 'undefined' && Config.DEBUG && typeof console !== 'undefined' && console.log) console.log(`Removed ${levelData.hazards.length} hazards from level load (global hazard removal).`);
    }

    /**
     * Update animated level elements (moving platforms, hazards).
     * @param {number} deltaTime - Time (seconds) since last update
     */
    update(deltaTime) {
        // Update moving platforms
        const time = Date.now() / 1000;
        
        this.platforms.forEach(plat => {
            if (plat.type === 'moving') {
                // Simple Sine wave movement
                const range = plat.range || 100;
                const speed = plat.speed || 1;
                
                if (plat.axis === 'y') {
                    plat.y = plat.initialY + Math.sin(time * speed + plat.timeOffset) * range;
                    // Store velocity to move player with platform later if needed
                    plat.dy = (Math.cos(time * speed + plat.timeOffset) * range * speed); 
                } else {
                    plat.x = plat.initialX + Math.sin(time * speed + plat.timeOffset) * range;
                    plat.dx = (Math.cos(time * speed + plat.timeOffset) * range * speed);
                }
            }
        });

        // Update moving hazards (if any)
        if (this.hazards && this.hazards.length > 0) {
            this.hazards.forEach(h => {
                // Generic moving hazard support: if axis and range are present, animate position
                if (h.axis && h.range) {
                    const speed = h.speed || 1.0;
                    if (!('initialX' in h)) h.initialX = h.x;
                    if (!('initialY' in h)) h.initialY = h.y;
                    if (!('timeOffset' in h)) h.timeOffset = Math.random() * Math.PI * 2;
                    if (h.axis === 'y') {
                        h.y = h.initialY + Math.sin(time * speed + h.timeOffset) * h.range;
                    } else {
                        h.x = h.initialX + Math.sin(time * speed + h.timeOffset) * h.range;
                    }
                }
            });
        }
    }

    /**
     * Check collision against platforms using previous-frame position to avoid tunneling.
     * Returns { collided: boolean, platform?: Object, landingY?: number }
     * @param {Object} rect - Current object bounding rect { x,y,width,height }
     * @param {Object} prevRect - Previous frame bounding rect
     * @param {number} velocityY - Current vertical velocity (used to skip upward movement)
     */
    checkPlatformCollision(rect, prevRect, velocityY) {
        // Optimization: Don't check if moving up
        if (velocityY < 0) return { collided: false };

        const rectBottom = rect.y + rect.height;
        const prevBottom = prevRect.y + prevRect.height;

        for (const platform of this.platforms) {
            // 1. Horizontal Overlap Check
            if (rect.x + rect.width > platform.x && rect.x < platform.x + platform.width) {
                
                // 2. Vertical "Crossed the Line" Check
                // Did we exist ABOVE the platform in the last frame?
                const wasAbove = prevBottom <= platform.y;
                // Are we BELOW (or ON) the platform in this frame?
                const isBelow = rectBottom >= platform.y;

                // If we crossed the threshold, it's a collision
                if (wasAbove && isBelow) {
                    return {
                        collided: true,
                        platform: platform,
                        // Snap exactly to top of platform
                        landingY: platform.y - rect.height 
                    };
                }
            }
        }
        return { collided: false };
    }

    /**
     * Render the level: background layers, static layer, moving platforms.
     * @param {CanvasRenderingContext2D} ctx
     * @param {number} [cameraX=0]
     * @param {number} [cameraY=0]
     * @param {number|null} [viewWidth=null]
     * @param {number|null} [viewHeight=null]
     */
    draw(ctx, cameraX = 0, cameraY = 0, viewWidth = null, viewHeight = null) {
        // 1. Draw Background (screen-space): panorama if available, otherwise gradient.
        // Important: background should fill the viewport (0..w,0..h) regardless of camera.
        const w = viewWidth || this.width || ctx.canvas.width;
        const h = viewHeight || this.height || ctx.canvas.height;

        // Mobile performance tuning: allow the same draw path, but optionally
        // disable heavy background images on very low-end devices.
        let mobilePerfMode = null;
        try { mobilePerfMode = (typeof localStorage !== 'undefined') ? localStorage.getItem('mobilePerfMode') : null; } catch (e) {}
        const allowBackgroundImage = !(this.useMobileOptimizations && mobilePerfMode === 'low');

        let bgImg = null;
        // Always attempt to draw the main background image when available
        // unless explicitly disabled by low-end mobile perf mode.
        if (allowBackgroundImage) {
            try {
                if (!this.cachedSprites[this.backgroundName]) {
                    this.cachedSprites[this.backgroundName] = (typeof spriteLoader !== 'undefined') ? spriteLoader.getSprite(this.backgroundName) : null;
                }
                bgImg = this.cachedSprites[this.backgroundName];
            } catch (e) { bgImg = null; }
        }

        // Main background first (farthest layer)
        if (bgImg) {
            try {
                const scaleY = h / bgImg.height;
                const tileW = Math.max(1, Math.ceil(bgImg.width * scaleY));
                const parallax = (typeof this.backgroundParallax !== 'undefined') ? this.backgroundParallax : (typeof Config !== 'undefined' ? Config.BACKGROUND_PARALLAX : 0.5);
                const repeatCount = Math.ceil(w / tileW) + 2;
                const startX = Math.floor(-((cameraX * parallax) % tileW));
                for (let i = 0; i < repeatCount; i++) {
                    const dx = startX + i * tileW;
                    ctx.drawImage(bgImg, 0, 0, bgImg.width, bgImg.height, dx, 0, tileW, h);
                }
            } catch (e) {
                bgImg = null;
            }
        }

        // Fallback gradient if background image isn't available
        if (!bgImg) {
            // Recreate gradient if viewport height changes
            if (!this.backgroundGradient || this._backgroundGradientH !== h) {
                this._backgroundGradientH = h;
                this.backgroundGradient = ctx.createLinearGradient(0, 0, 0, h);
                this.backgroundGradient.addColorStop(0, this.theme.bgTop);
                this.backgroundGradient.addColorStop(0.5, this.theme.bgMid);
                this.backgroundGradient.addColorStop(1, this.theme.bgBot);
            }
            ctx.fillStyle = this.backgroundGradient;
            ctx.fillRect(0, 0, w, h);
        }

        // Background layers (drawn on top of main background for depth)
        if (!this.useMobileOptimizations) {
            for (const layer of this.backgroundLayers) {
                let layerImg = null;
                try {
                    if (!layer || !layer.name) continue;
                    if (!this.cachedSprites[layer.name]) {
                        this.cachedSprites[layer.name] = (typeof spriteLoader !== 'undefined') ? spriteLoader.getSprite(layer.name) : null;
                    }
                    layerImg = this.cachedSprites[layer.name];
                } catch (e) { layerImg = null; }
                if (!layerImg) continue;

                try {
                    const scaleY = h / layerImg.height;
                    const tileW = Math.max(1, Math.ceil(layerImg.width * scaleY));
                    const parallax = (typeof layer.parallax === 'number') ? layer.parallax : 0.5;
                    const repeatCount = Math.ceil(w / tileW) + 2;
                    const startX = Math.floor(-((cameraX * parallax) % tileW));
                    for (let i = 0; i < repeatCount; i++) {
                        const dx = startX + i * tileW;
                        ctx.drawImage(layerImg, 0, 0, layerImg.width, layerImg.height, dx, 0, tileW, h);
                    }
                } catch (e) {}
            }
        }

        ctx.save();
        ctx.translate(-cameraX, -cameraY);

        // 2. Draw Platforms — use pre-rendered static layer for non-moving
        // platforms to speed up rendering on low-end devices.
        if (this._staticLayerCanvas && !this._staticNeedsUpdate) {
            try {
                ctx.imageSmoothingEnabled = false;
                // Draw entire static layer stretched to the level size.
                ctx.drawImage(this._staticLayerCanvas, 0, 0, this._staticLayerCanvas.width, this._staticLayerCanvas.height, 0, 0, this.width, this.height);
            } catch (e) {
                // fallback to per-platform draw
                for (const platform of this.platforms) {
                    if (platform.type !== 'moving') this.drawPlatform(ctx, platform);
                }
            }
        } else {
            for (const platform of this.platforms) {
                if (platform.type !== 'moving') this.drawPlatform(ctx, platform);
            }
        }

        // Always draw moving platforms on top
        for (const platform of this.platforms) {
            if (platform.type === 'moving') this.drawPlatform(ctx, platform);
        }

        // 3. Hazards are not supported in this build; legacy hazard data is ignored.

        ctx.restore();
    }

    constructor(width = 800, height = 600) {
        this.width = width;
        this.height = height;
        this.platforms = [];

        // Level visuals & content
        this.backgroundName = 'bg_city';
        this.backgroundLayers = [];
        this.backgroundParallax = (typeof Config !== 'undefined' ? Config.BACKGROUND_PARALLAX : 0.5);
        this.spawnPoints = null;
        this.hazards = [];

        this.backgroundGradient = null;
        this.tileMode = 'tiles'; // 'tiles' or 'neon' - controls platform rendering
        this._tilePatterns = {}; // reserved for future per-canvas pattern caching
        this.cachedSprites = {}; // cache for sprite images

        // Static layer caching for pre-rendered non-moving platforms
        this._staticNeedsUpdate = true;
        this._staticLayerCanvas = null;
        this._staticLayerScale = 1;

        // Performance flags
        this.useMobileOptimizations = false;

        // Cyberpunk style config
        this.theme = {
            // Default/fallback background: light pastel green gradient
            // (used when no panorama is available or when low-end mobile perf mode disables it)
            bgTop: '#eefaf0',
            bgMid: '#d6f3da',
            bgBot: '#bfecc6',
            platTop: '#00fff7',    // Neon cyan
            platBot: '#ff00ea',    // Neon magenta
            border: '#fffb00',     // Bright yellow border
            glow: '#00fff7',       // Neon cyan glow
        };
    }

    /**
     * Load and cache a sprite by name using the global spriteLoader.
     * Returns null if the sprite isn't available.
     * @param {string} name
     * @returns {HTMLImageElement|null}
     */
    _getSprite(name) {
        if (!name) return null;
        if (!this.cachedSprites[name]) {
            try {
                this.cachedSprites[name] = (typeof spriteLoader !== 'undefined') ? spriteLoader.getSprite(name) : null;
            } catch (e) {
                this.cachedSprites[name] = null;
                if (typeof console !== 'undefined' && console.warn) console.warn(`Failed to load sprite ${name}`, e);
            }
        }
        return this.cachedSprites[name];
    }

    /**
     * Create a repeating CanvasPattern for the given tile on the provided context.
     * Returns null on failure.
     * @param {CanvasRenderingContext2D} ctx
     * @param {string} tileName
     * @returns {CanvasPattern|null}
     */
    _createPattern(ctx, tileName) {
        if (!ctx || !tileName) return null;
        const tileImg = this._getSprite(tileName);
        if (!tileImg) return null;
        try {
            return ctx.createPattern(tileImg, 'repeat');
        } catch (e) {
            if (typeof console !== 'undefined' && console.warn) console.warn(`Failed to create pattern for ${tileName}`, e);
            return null;
        }
    }

    drawPlatform(ctx, p) {
        // If tile mode is enabled and sprite exists, draw a tiled fill
        if (this.tileMode === 'tiles') {
            const tileName = p.tile || 'platform_tile';
            const pattern = this._createPattern(ctx, tileName);
            if (pattern) {
                ctx.save();
                ctx.fillStyle = pattern;
                ctx.fillRect(p.x, p.y, p.width, p.height);
                ctx.restore();

                // Subtle border and highlight for readability
                ctx.save();
                ctx.strokeStyle = 'rgba(0,0,0,0.6)';
                ctx.lineWidth = 2;
                ctx.strokeRect(p.x, p.y, p.width, p.height);
                ctx.globalAlpha = 0.25;
                ctx.fillStyle = '#fff';
                ctx.fillRect(p.x, p.y, p.width, 4);
                ctx.restore();

                return; // done
            }
        }

            // Fallback: Cyberpunk neon shadow/glow (reduced on mobile for perf)
            ctx.save();
            ctx.shadowColor = this.theme.glow;
            const glowBlur = this.useMobileOptimizations ? 6 : 20;
            ctx.shadowBlur = glowBlur;
            ctx.fillStyle = this.useMobileOptimizations ? 'rgba(0,255,247,0.08)' : 'rgba(0,255,247,0.15)';
            ctx.fillRect(p.x - 8, p.y - 8, p.width + 16, p.height + 16);
            ctx.restore();

           // Neon gradient fill
           const grad = ctx.createLinearGradient(p.x, p.y, p.x, p.y + p.height);
           grad.addColorStop(0, this.theme.platTop);
           grad.addColorStop(1, this.theme.platBot);
           ctx.fillStyle = grad;
           ctx.fillRect(p.x, p.y, p.width, p.height);

           // Neon border
           ctx.save();
           ctx.shadowColor = this.theme.border;
           const borderBlur = this.useMobileOptimizations ? 3 : 10;
           ctx.shadowBlur = borderBlur;
           ctx.strokeStyle = this.theme.border;
           ctx.lineWidth = 3;
           ctx.strokeRect(p.x, p.y, p.width, p.height);
           ctx.restore();

           // Neon highlight
           ctx.save();
           ctx.globalAlpha = 0.25;
           ctx.fillStyle = '#fff';
           ctx.fillRect(p.x, p.y, p.width, 6);
           ctx.restore();
    }

    // ----- Static layer pre-rendering for non-moving platforms -----
    // Call when level content or viewport size changes
    renderStaticLayer(viewWidth = null, viewHeight = null) {
        try {
            // Cap pre-rendered static layer width to avoid excessive memory
            // and GPU texture usage. Use a smaller cap on mobile to reduce
            // pressure on devices like iPad/Safari.
            const MAX_STATIC_WIDTH = this.useMobileOptimizations ? 2048 : 4096;
            const targetW = Math.min(this.width, MAX_STATIC_WIDTH);
            // We'll render at the level height so vertical content is preserved
            const targetH = Math.max(1, Math.floor(this.height));

            const canvas = document.createElement('canvas');
            canvas.width = Math.max(1, Math.floor(targetW));
            canvas.height = Math.max(1, Math.floor(targetH));
            const c = canvas.getContext('2d');
            c.imageSmoothingEnabled = false;

            // Clear
            c.fillStyle = 'rgba(0,0,0,0)';
            c.fillRect(0, 0, canvas.width, canvas.height);

            // Draw static (non-moving) platforms into the static canvas
            for (const p of this.platforms) {
                if (p.type === 'moving') continue; // skip moving platforms

                const sx = Math.floor((p.x / this.width) * canvas.width);
                const sy = Math.floor((p.y / this.height) * canvas.height);
                const sw = Math.max(1, Math.floor((p.width / this.width) * canvas.width));
                const sh = Math.max(1, Math.floor((p.height / this.height) * canvas.height));

                // If tile mode with a pattern, create pattern on the static ctx
                if (this.tileMode === 'tiles') {
                    const tileName = p.tile || 'platform_tile';
                    const pattern = this._createPattern(c, tileName);
                    if (pattern) {
                        c.save();
                        c.fillStyle = pattern;
                        c.fillRect(sx, sy, sw, sh);
                        c.restore();
                        // subtle border
                        c.save(); c.strokeStyle = 'rgba(0,0,0,0.6)'; c.lineWidth = 2; c.strokeRect(sx, sy, sw, sh); c.restore();
                        continue;
                    }
                }

                // Fallback neon style
                c.save();
                const grad = c.createLinearGradient(sx, sy, sx, sy + sh);
                grad.addColorStop(0, this.theme.platTop);
                grad.addColorStop(1, this.theme.platBot);
                c.fillStyle = grad;
                c.fillRect(sx, sy, sw, sh);
                c.restore();
            }

            this._staticLayerCanvas = canvas;
            this._staticLayerScale = canvas.width / this.width;
            this._staticNeedsUpdate = false;
        } catch (e) {
            this._staticLayerCanvas = null;
            this._staticNeedsUpdate = true;
            console.warn('renderStaticLayer failed', e);
        }
    }
}